File: src/listener.py
#!/usr/bin/env python
import rospy
from std_msgs.msg import String

def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.data)


def listener():

    rospy.init_node('listener', anonymous=True)

    rospy.Subscriber("chatter", String, callback)
    # This declares that your node subscribes to the chatter topic which is of type std_msgs.msgs.String.
     # When new messages are received, callback is invoked with the message as the first argument. 

    rospy.spin()


if __name__ == '__main__':
    listener()

File: src/listener.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"

// Topic messages callback
// this only takes one parameter which is (msgs)
// (ConstPtr&) is a pointer for the message
void chatterCallback(const std_msgs::String::ConstPtr& msg)
{
  ROS_INFO("I heard: [%s]", msg->data.c_str());
}

int main(int argc, char **argv)
{
  //  initiate a new ros node named "listener"
  ros::init(argc, argv, "listener");

  // create a node handle: it is reference assigned to a new node
  ros::NodeHandle node;  //you can give this any name

  // Subscribe to a given topic, in this case "chatter"
  //chatter callback is the name of the callback funciton that will be executed each time a message is received
  ros::Subscriber sub = node.subscribe("chatter", 1000, chatterCallback); //(<name of topic>, <buffer size>, <fucntion which is a callback)

  // enter a lopp, pumping callbacks
  ros::spin();

  return 0;

}
